core/modules/mlfqueue.cc

```diff
@@ -0,0 +1,346 @@
+#include "mlfqueue.h"
+#include "../mem_alloc.h"
+
+#include <rte_ether.h>
+#include <rte_ip.h>
+#include <rte_udp.h>
+#include <time.h>
+
+#include <glog/logging.h>
+
+#include <string>
+#include <cmath>
+#include <iostream>
+#include <fstream>
+
+bool operator==(const MLFQueue::Q_Id &id1, const MLFQueue::Q_Id &id2) {
+  bool ips = id1.src_ip == id2.src_ip && id1.dst_ip == id2.dst_ip;
+  bool ports = id1.src_port == id2.src_port && id1.dst_port == id2.dst_port;
+  return ips && ports && id1.protocol == id2.protocol;
+}
+bool operator<(const MLFQueue::Q_Id &id1, const MLFQueue::Q_Id &id2)  {
+  bool ips = id1.src_ip < id2.src_ip && id1.dst_ip < id2.dst_ip;
+  bool ports = id1.src_port < id2.src_port && id1.dst_port < id2.dst_port;
+  return ips && ports && id1.protocol < id2.protocol;
+}
+
+int RoundToPowerTwo(int val) {
+ return pow(2, ceil(log2(val)));
+}
+
+const Commands MLFQueue::cmds = {
+    {"set_num_lvls", "MlfqLvlArg", MODULE_CMD_FUNC(&MLFQueue::CommandNumPriorityLevels), 0},
+    {"set_batch_Size", "MlfqBatchArg", MODULE_CMD_FUNC(&MLFQueue::CommandBatchSize), 0},
+    {"set_max_flow_queue_size", "MlfqMaxFlowQueueSizeArg",
+          MODULE_CMD_FUNC(&MLFQueue::CommandMaxFlowQueueSize), 0}};
+
+pb_error_t MLFQueue::Init(const bess::pb::MlfqArg &arg) {
+  pb_error_t err;
+
+  if (arg.num_lvls() != 0) {
+    err = SetNumPriorityLevels(arg.num_lvls());
+    if (err.err() != 0) {
+      return err;
+    }
+  } else {
+    err = SetNumPriorityLevels(DEFAULT_NUM_LVLS);
+    if (err.err() != 0) {
+      return err;
+    }
+  }
+
+  if (arg.max_flow_queue_size() != 0) {
+    err = SetMaxFlowQueueSize(arg.max_flow_queue_size());
+    if (err.err() != 0) {
+      return err;
+    }
+  } else {
+    err = SetMaxFlowQueueSize(FLOW_QUEUE_MAX);
+    if (err.err() != 0) {
+      return err;
+    }
+  }
+
+  if (arg.batch_size() != 0) {
+    err = SetBatchSize(arg.batch_size());
+    if (err.err() != 0) {
+      return err;
+    }
+  } else {
+    err = SetBatchSize(DEFAULT_BATCH_SIZE);
+    if (err.err() != 0) {
+      return err;
+    }
+  }
+
+  if(arg.init_load() != 0) {
+    load_avg_ = arg.init_load();
+  } else {
+    load_avg_ = INITIAL_LOAD;
+  }
+  ready_flows_ = 0;
+  return pb_errno(0);
+}
+
+void MLFQueue::DeInit() {
+  bess::Packet *pkt;
+
+  for (std::map<Q_Id, Flow>::iterator it=flows_.begin(); it!=flows_.end(); ) {
```

Comment:
Unless I'm missing it, I don't see where you advance the iterator here.

Also, prefer auto for iterator types to reduce clutter.